### PR TITLE
[REVIEW] detail::scatter variant with map iterator support

### LINKED ISSUE #3745: [FEA] scatter should have a detail API that accepts a scatter map iterator
**Is your feature request related to a problem? Please describe.**

`gather` exposes a [`detail` API](https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/include/cudf/detail/gather.cuh#L302) that accepts an iterator for the gather map. 

`scatter` should have the same thing, but it currently does not. 

**Describe the solution you'd like**

A `cudf/detail/scatter.cuh` file should be created with a `scatter` API that accepts an iterator for the scatter map. The existing `scatter` implementation should then be updated to use this internal API. 

This should mirror the structure of gather as much as possible for consistency. 

It does appear `scatter`'s internal implementation does use an[ iterator for the map internally](https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/src/copying/scatter.cu#L100), but this needs to be exposed in a detail API.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Adding detail::scatter variant with map iterator.

closes #3745

### CODE REVIEW

File: cpp/include/cudf/detail/scatter.cuh [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+ * @param[in] mr The resource to use for all allocations
+ * @param[in] stream The stream to use for CUDA operations
+ *
+ * @return Result of scattering values from source to target
+ **/
+template <typename T, typename MapIterator>
+std::unique_ptr<table>
+scatter(table_view const& source, MapIterator scatter_map_begin,
+    MapIterator scatter_map_end, table_view const& target,
+    bool check_bounds = false,
```
Comment (devavret) [inline cpp/include/cudf/detail/scatter.cuh]: This doesn't do anything. The API shouldn't accept this argument if it does nothing with it. Alternatively, you can pull out the bounds checking code, make it a separate function and call it from inside this one.

File: cpp/include/cudf/detail/scatter.cuh [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  template <typename T, std::enable_if_t<is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& source,
+      MapIterator scatter_map_begin, MapIterator scatter_map_end, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    auto result = std::make_unique<column>(target, stream, mr);
+    auto result_view = result->mutable_view();
+
+    // NOTE use source.begin + scatter rows rather than source.end in case the
+    // scatter map is smaller than the number of source rows
```
Comment (devavret) [inline cpp/include/cudf/detail/scatter.cuh]: This comment is the opposite of the behaviour immediately following it. This indicates the source size is limiting whereas the code says iterator distance is limiting.

Comment (rgsl888prabhu) [reply cpp/include/cudf/detail/scatter.cuh]:
```
thrust::scatter(rmm::exec_policy(stream)->on(stream), source.begin<T>(),
      source.begin<T>() + std::distance(scatter_map_begin, scatter_map_end),
      scatter_map_begin,
      result_view.begin<T>());
```
I am limiting the source size to size of the scatter map. Do you want to have two conditions where one limited by source size if source.size() < map.size() and other one which already exists?

Comment (devavret) [reply cpp/include/cudf/detail/scatter.cuh]: I misunderstood something.

File: cpp/include/cudf/detail/gather.cuh [hunk 3]
```diff
@@ -1,3 +1,20 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
```
Comment (trevorsm7) [inline cpp/include/cudf/detail/gather.cuh]:
```suggestion
 * Copyright (c) 2019-2020, NVIDIA CORPORATION.
```

File: cpp/include/cudf/detail/gather.cuh [hunk 4]
```diff
@@ -1,3 +1,20 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
 #include <cudf/cudf.h>
```
Comment (trevorsm7) [inline cpp/include/cudf/detail/gather.cuh]: Remove this header if possible
```suggestion
```

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Pull request merged**